src/qibolab/instruments/zhinst/executor.py

```diff
@@ -0,0 +1,781 @@
+                        ),
+                        measure_pulse_parameters=measure_pulse_parameters,
+                        measure_pulse_amplitude=None,
+                        acquire_delay=self.time_of_flight * NANO_TO_SECONDS,
+                        reset_delay=reset_delay,
+                    )
+            previous_section = section_uid
+
+    @staticmethod
+    def play_sweep(exp, channel_name, pulse):
+        """Play Zurich pulse when a single sweeper is involved."""
+        play_parameters = {}
+        for p, zhs in pulse.zhsweepers:
+            if p is Parameter.amplitude:
+                pulse.zhpulse.amplitude *= max(zhs.values)
+                zhs.values /= max(zhs.values)
+                play_parameters["amplitude"] = zhs
+            if p is Parameter.duration:
+                play_parameters["length"] = zhs
+            if p is Parameter.relative_phase:
+                play_parameters["phase"] = zhs
+        if "phase" not in play_parameters:
+            play_parameters["phase"] = pulse.pulse.relative_phase
+
+        exp.play(signal=channel_name, pulse=pulse.zhpulse, **play_parameters)
+
+    @staticmethod
+    def rearrange_rt_sweepers(
+        sweepers: list[Sweeper],
+    ) -> tuple[np.ndarray, list[Sweeper]]:
+        """Rearranges sweepers from qibocal based on device hardware
+        limitations.
+
+        The only known limitation currently is that frequency sweepers must be applied before (on the outer loop) other
+        (e.g. amplitude) sweepers. Consequently, the only thing done here is to swap the frequency sweeper with the
+        first sweeper in the list.
+
+        Args:
+            sweepers (list[Sweeper]): Sweepers to rearrange.
+
+        Returns:
+            swapped_axis_pair (np.ndarray): array of shape (2,) containing the indices of the two swapped axes.
+            sweepers (list[Sweeper]): rearranged list of sweepers.
+        """
+        swapped_axis_pair = np.zeros(2, dtype=int)
+        sweepers_copy = sweepers.copy()
+        freq_sweeper = next(
+            iter(s for s in sweepers_copy if s.parameter is Parameter.frequency), None
+        )
+        if freq_sweeper:
+            freq_sweeper_idx = sweepers_copy.index(freq_sweeper)
+            sweepers_copy[freq_sweeper_idx] = sweepers_copy[0]
+            sweepers_copy[0] = freq_sweeper
+            swapped_axis_pair = np.array([0, freq_sweeper_idx])
+            log.warning("Sweepers were reordered")
+        return swapped_axis_pair, sweepers_copy
+
+    def sweep(self, qubits, couplers, sequence: PulseSequence, options, *sweepers):
+        """Play pulse and sweepers sequence."""
+
+        self.signal_map = {}
+        self.processed_sweeps = ProcessedSweeps(sweepers, qubits)
+        self.nt_sweeps, self.rt_sweeps = classify_sweepers(sweepers)
+        swapped_axis_pair, self.rt_sweeps = self.rearrange_rt_sweepers(self.rt_sweeps)
+        swapped_axis_pair += len(self.nt_sweeps)
```

Comment:
I don't need them to be attributes. I just kept them as attributes so that it remains similar to the original implementation. I would eventually reduce state maintenance of the object close to zero, but maybe later.

Reply:
Btw, what do you mean by _reduce state maintenance of the object close to zero_ ?

Reply:
The _state_ is composed by the data that are associated to the object itself, so the content of `self`. The problem caused by having a wide amount of data in `self` is isolation: when things are happening, you hardly know who set the value you're looking at (as it is happening, even more, with global variables).

Instead, function arguments are local, you know where they come from. It's in general easier to debug with only local information.